### PR TITLE
Only override objects from informer when version has increased.

### LINKED ISSUE #63467: TestVolumeBindingStress still flaking
**Is this a BUG REPORT or FEATURE REQUEST?**:
@k8s-sig-storage-bugs
cc @msau42 

**What happened**:
VolumeBindingStress integration test started flaking around 4/26.  Originally thought it was due to #62929, but even after reverting, the test is still flaking.
logs: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/63045/pull-kubernetes-integration/9817/
testgrid: https://k8s-testgrid.appspot.com/presubmits-kubernetes-blocking#pull-kubernetes-integration

### PR DESCRIPTION
**Pull request by msau42**
**What this PR does / why we need it**:
We don't want an informer resync to override assumed volumes if the version has not increased.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #63467

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/scheduler_assume_cache.go [hunk 1]
```diff
@@ -158,8 +158,29 @@ func (c *assumeCache) add(obj interface{}) {
...
+
+		storedVersion, err := c.getObjVersion(name, objInfo.latestObj)
+		if err != nil {
+			glog.Errorf("add: couldn't get stored object version: %v", err)
+			return
+		}
+
+		// Only update object if version is newer.
+		// This is so we don't override assumed objects due to informer resync.
+		if newVersion <= storedVersion {
```
Comment (wgliang) [inline pkg/controller/volume/persistentvolume/scheduler_assume_cache.go]: It would be better to add a log here.

Comment (msau42) [reply pkg/controller/volume/persistentvolume/scheduler_assume_cache.go]: done

**Comment by msau42** [comment]
/hold
Need to test with "-race" option

**Comment by msau42** [comment]
/unhold

I reran the tests with -race option, which causes the tests to run twice as long.  So I had to increase the timeout on the test.

File: pkg/controller/volume/persistentvolume/scheduler_assume_cache_test.go [hunk 2]
```diff
@@ -243,6 +243,39 @@ func TestPVCacheWithStorageClasses(t *testing.T) {
...
+	if err := cache.Assume(newPV); err != nil {
+		t.Fatalf("failed to assume PV: %v", err)
+	}
+	if err := getPV(cache, pvName, newPV); err != nil {
+		t.Fatalf("failed to get PV after assume: %v", err)
+	}
+
+	// Add old PV
+	internal_cache.add(pv)
+	if err := getPV(cache, pvName, newPV); err != nil {
```
Comment (verult) [inline pkg/controller/volume/persistentvolume/scheduler_assume_cache_test.go]: nit: "`verifyPV`"?

Comment (msau42) [reply pkg/controller/volume/persistentvolume/scheduler_assume_cache_test.go]: updated

File: pkg/controller/volume/persistentvolume/scheduler_binder_test.go [hunk 3]
```diff
@@ -734,3 +740,68 @@ func TestBindPodVolumes(t *testing.T) {
...
+		t.Errorf("Test failed: binding not required")
+	}
+	testEnv.validateAssume(t, "assume", pod, expectedBindings)
+	// After assume, claimref should be set on pv
+	expectedBindings = testEnv.getPodBindings(t, "after-assume", testNode.Name, pod)
+
+	// 3. Find matching PVs again
+	// This should always return the original chosen pv
+	t.Logf("Testing FindPodVolumes after Assume")
+	for i := 0; i < 50; i++ {
```
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/controller/volume/persistentvolume/scheduler_binder_test.go]: added a comment

File: test/integration/scheduler/volume_binding_test.go [hunk 4]
```diff
@@ -288,7 +288,9 @@ func TestVolumeBindingStress(t *testing.T) {
 
 	// Validate Pods scheduled
 	for _, pod := range pods {
-		if err := waitForPodToSchedule(config.client, pod); err != nil {
+		// Use increased timeout for stress test because there is a higher chance of
+		// PV sync error
+		if err := waitForPodToScheduleWithTimeout(config.client, pod, 60*time.Minute); err != nil {
```
Comment (resouer) [inline test/integration/scheduler/volume_binding_test.go]: Per discussion with @msau42 on slack and 2~hrs non-stop testing locally,  I can confirm the issue is fixed. Thanks!

Emm .. do we really need `60*time.Minute` timeout here? I recalled the overall timeout of integration is 300s.

Comment (msau42) [reply test/integration/scheduler/volume_binding_test.go]: Yeah I ran this test over 1000 times until all the log files filled up my filesystem :-D

I totally meant 60 seconds, not minutes! Thanks for catching

**Comment by resouer** [comment]
/lgtm

cc @bsalamat for final pass

File: pkg/controller/volume/persistentvolume/scheduler_assume_cache.go [hunk 5]
```diff
@@ -158,8 +158,30 @@ func (c *assumeCache) add(obj interface{}) {
...
```
Comment (k82cn) [inline pkg/controller/volume/persistentvolume/scheduler_assume_cache.go]: Just curious why no issue in other place? AFAIK, most of code did not handle that.

Comment (msau42) [reply pkg/controller/volume/persistentvolume/scheduler_assume_cache.go]: What other places are you referring to?

**Review by bsalamat — COMMENTED** [review]
/lgtm
for the scheduler changes

**Pull request merged**